torchtext/experimental/datasets/translation.py

```diff
@@ -0,0 +1,388 @@
+    datasets = []
+    for key in data_select:
+        src_text_transform = sequential_transforms(src_text_vocab_transform,
+                                                   vocab_func(src_vocab),
+                                                   totensor(dtype=torch.long))
+        tgt_text_transform = sequential_transforms(tgt_text_vocab_transform,
+                                                   vocab_func(tgt_vocab),
+                                                   totensor(dtype=torch.long))
+        datasets.append(
+            TranslationDataset(raw_data[key], (src_vocab, tgt_vocab),
+                               (src_text_transform, tgt_text_transform)))
+
+    return tuple(datasets)
+
+
+class TranslationDataset(torch.utils.data.Dataset):
+    """Defines a dataset for translation.
+       Currently, we only support the following datasets:
+             - Multi30k
+             - WMT14
+             - IWSLT
+    """
+    def __init__(self, data, vocab, transforms):
+        """Initiate translation dataset.
+
+        Arguments:
+            data: a tuple of source and target tensors, which include token ids
+                numericalizing the string tokens.
+                [(src_tensor0, tgt_tensor0), (src_tensor1, tgt_tensor1)]
+            vocab: source and target Vocabulary object used for dataset.
+                (src_vocab, tgt_vocab)
+            transforms: a tuple of source and target string transforms.
+
+        Examples:
+            >>> from torchtext.vocab import build_vocab_from_iterator
+            >>> src_data = torch.Tensor([token_id_s1, token_id_s2,
+                                         token_id_s3, token_id_s1]).long()
+            >>> tgt_data = torch.Tensor([token_id_t1, token_id_t2,
+                                         token_id_t3, token_id_t1]).long()
+            >>> src_vocab = build_vocab_from_iterator([['Übersetzungsdatensatz']])
+            >>> tgt_vocab = build_vocab_from_iterator([['translation', 'dataset']])
+            >>> dataset = TranslationDataset([(src_data, tgt_data)],
+                                              (src_vocab, tgt_vocab))
+        """
+
+        super(TranslationDataset, self).__init__()
+        self.data = data
+        self.vocab = vocab
+        self.transforms = transforms
+
+    def __getitem__(self, i):
+        source = self.transforms[0](self.data[i][0])
+        target = self.transforms[1](self.data[i][1])
+        return (source, target)
+
+    def __len__(self):
+        return len(self.data)
+
+    def get_vocab(self):
+        return self.vocab
+
+
+def Multi30k(languages="de-en",
+             train_filename="train",
+             valid_filename="val",
+             test_filename="test_2016_flickr",
+             tokenizer=None,
+             root='.data',
+             vocab=(None, None),
+             removed_tokens=['<unk>']):
+    """ Define translation datasets: Multi30k
+        Separately returns train/valid/test datasets as a tuple
```

Comment:
I think we should keep a copy of the available language/train_filename here because most users will the docs here.

Reply:
done